File: js/src/contracts.test.ts
import { readFileSync } from "node:fs";

import { describe, expect, expectTypeOf, it } from "vitest";

import {
  embedOne,
  embedTexts,
  isEmbeddingProvider,
  NotAnEmbedderError,
  type Awaitable,
  type EmbeddingProvider,
  type GeneratorProvider,
  type SingleTextEmbedder,
  type Vector,
} from "./contracts.js";
import * as root from "./index.js";

// ---------------------------------------------------------------------------
// Providers written the way an outsider would write them: object literals and
// plain classes that extend nothing of ours.
// ---------------------------------------------------------------------------

const syncEmbedder = {
  calls: [] as (readonly string[])[],
  embed(texts: readonly string[]): Vector[] {
    this.calls.push([...texts]);
    return texts.map((t) => [t.length, 1, 0]);
  },
};

const asyncEmbedder = {
  calls: [] as (readonly string[])[],
  async embed(texts: readonly string[]): Promise<Vector[]> {
    this.calls.push([...texts]);
    return texts.map((t) => [t.length, 1, 0]);
  },
};

const singleText: SingleTextEmbedder = (text: string) => [text.length, 1, 0];

class OutsideGenerator {
  seen: string[] = [];
  async answer(question: string, passage: string, answerLanguage = "en"): Promise<string> {
    this.seen.push(`${question}|${answerLanguage}`);
    return passage;
  }
}

describe("the published contracts (ADR-0014 R4)", () => {
  it("re-exports both contracts and their helpers from the package root", () => {
    expect(typeof root.embedTexts).toBe("function");
    expect(typeof root.embedOne).toBe("function");
    expect(typeof root.isEmbeddingProvider).toBe("function");
    expect(typeof root.singleFrom).toBe("function");
    expect(root.NotAnEmbedderError).toBe(NotAnEmbedderError);
  });

  it("is satisfied by shape alone — nothing extends a CiteNexus class", () => {
    expectTypeOf(syncEmbedder).toMatchTypeOf<EmbeddingProvider>();
    expectTypeOf(asyncEmbedder).toMatchTypeOf<EmbeddingProvider>();
    expectTypeOf(new OutsideGenerator()).toMatchTypeOf<GeneratorProvider>();
    expect(Object.getPrototypeOf(syncEmbedder)).toBe(Object.prototype);
  });

  it("accepts a synchronous AND an asynchronous provider (Awaitable<T>)", () => {
    // The strict `=> Promise<T>` spelling would lock the hermetic fakes — the
    // reference implementations of these very contracts — out of the contract.
    expectTypeOf<Awaitable<string>>().toEqualTypeOf<string | Promise<string>>();
    const sync: EmbeddingProvider = { embed: (t) => t.map(() => [1]) };
    const async_: EmbeddingProvider = { embed: async (t) => t.map(() => [1]) };
    expect(isEmbeddingProvider(sync) && isEmbeddingProvider(async_)).toBe(true);
  });

  it("publishes NO contract for the seams this port cannot consume", () => {
    // Completion, vision and reranking have no call site anywhere in the JS
    // port — `grep -i rerank js/src` is empty, and there is no Candidate type
    // a rerank contract could even be spelled in terms of. A contract nothing
    // consumes advertises support that does not exist. See design.md §1.
    const source = readFileSync(new URL("./contracts.ts", import.meta.url), "utf8");
    for (const absent of [
      "CompletionProvider",
      "VisionProvider",
      "RerankerProvider",
      "Candidate",
      "complete(",
      "describe(",
      "rerank(",
    ]) {
      expect(source).not.toContain(absent);
    }
  });

  it("names no transport concern in any contract method (R3)", () => {
    const source = readFileSync(new URL("./contracts.ts", import.meta.url), "utf8");
    const interfaces = source.slice(source.indexOf("export interface"));
    for (const banned of ["base_url", "baseUrl", "headers", "transport", "timeout", "apiKey"]) {
      expect(interfaces).not.toContain(`${banned}:`);
    }
  });
});

// ---------------------------------------------------------------------------
// The Python naming hazard, checked rather than assumed
// ---------------------------------------------------------------------------

describe("the single-text and batch shapes cannot be confused", () => {
  // Python had to name the batch method `embed_many` because `str` IS a
  // `Sequence[str]`. In TS the two seams are not even the same KIND of value:
  // the single-text seam is a FUNCTION, the batch contract is an OBJECT with an
  // `embed` method — so `typeof` discriminates them with certainty and the
  // natural name `embed` is safe.
  it("the type guard tells them apart", () => {
    expect(isEmbeddingProvider(syncEmbedder)).toBe(true);
    expect(isEmbeddingProvider(asyncEmbedder)).toBe(true);
    expect(isEmbeddingProvider(singleText)).toBe(false);
    expect(isEmbeddingProvider(null)).toBe(false);
    expect(isEmbeddingProvider(undefined)).toBe(false);
    expect(isEmbeddingProvider({})).toBe(false);
    expect(isEmbeddingProvider({ embed: 3 })).toBe(false);
    expect(isEmbeddingProvider("embed")).toBe(false);
  });

  it("a string is not assignable where a batch of texts is required", () => {
    expectTypeOf<string>().not.toMatchTypeOf<readonly string[]>();
  });
});

// ---------------------------------------------------------------------------
// Dispatch
// ---------------------------------------------------------------------------

describe("embedTexts — the one place the port decides how to talk to an embedder", () => {
  it("prefers the batch contract and preserves input order", async () => {
    const provider = {
      batches: [] as string[][],
      embed(texts: readonly string[]): Vector[] {
        this.batches.push([...texts]);
        return texts.map((t) => [t.length]);
      },
    };
    const vecs = await embedTexts(provider, ["a", "bb", "ccc"]);
    expect(provider.batches).toEqual([["a", "bb", "ccc"]]);
    expect(vecs).toEqual([[1], [2], [3]]);
  });

  it("falls back to the deprecated single-text shape, per text, in order", async () => {
    const seen: string[] = [];
    const single: SingleTextEmbedder = (text) => {
      seen.push(text);
      return [text.length];
    };
    expect(await embedTexts(single, ["a", "bb"])).toEqual([[1], [2]]);
    expect(seen).toEqual(["a", "bb"]);
  });

  it("calls no model at all for an empty input", async () => {
    const provider = { batches: 0, embed(t: readonly string[]) { this.batches++; return t.map(() => [1]); } };
    expect(await embedTexts(provider, [])).toEqual([]);
    expect(provider.batches).toBe(0);
  });

  it("propagates a provider rejection unchanged — never a placeholder vector", async () => {
    const failing: EmbeddingProvider = {
      embed: () => Promise.reject(new Error("model call timed out after 30s")),
    };
    await expect(embedTexts(failing, ["a"])).rejects.toThrow(/timed out/);

    const failingSingle: SingleTextEmbedder = () => Promise.reject(new Error("timed out"));
    await expect(embedTexts(failingSingle, ["a"])).rejects.toThrow(/timed out/);
  });

  it("refuses a value that is neither shape, loudly", async () => {
    await expect(embedTexts({} as never, ["a"])).rejects.toThrow(NotAnEmbedderError);
    await expect(embedTexts(null as never, ["a"])).rejects.toThrow(NotAnEmbedderError);
  });

  it("refuses a batch whose length does not match the input", async () => {
    const short: EmbeddingProvider = { embed: () => [[1]] };
    await expect(embedTexts(short, ["a", "b"])).rejects.toThrow(/1 vectors for 2 texts/);
  });

  it("embedOne is a batch of one", async () => {
    const provider = {
      batches: [] as string[][],
      embed(texts: readonly string[]): Vector[] {
        this.batches.push([...texts]);
        return texts.map((t) => [t.length]);
      },
    };
    expect(await embedOne(provider, "hello")).toEqual([5]);
    expect(provider.batches).toEqual([["hello"]]);
  });
});

describe("singleFrom — a batch provider plugs into the single-text ingest seam", () => {
  it("adapts without the author writing glue", async () => {
    const adapted = root.singleFrom(syncEmbedder);
    expect(await adapted("hello")).toEqual([5, 1, 0]);
  });

  it("reports a short batch rather than returning a placeholder", async () => {
    await expect(root.singleFrom({ embed: () => [] })("hello")).rejects.toThrow();
  });
});

File: golang/contracts/contracts_test.go
package contracts

import (
	"errors"
	"go/ast"
	"go/parser"
	"go/token"
	"os"
	"path/filepath"
	"reflect"
	"strings"
	"testing"
)

// ---------------------------------------------------------------------------
// Providers written the way an outsider would write them. Nothing below embeds,
// extends, or otherwise names a CiteNexus concrete type.
// ---------------------------------------------------------------------------

type outsideBatchEmbedder struct {
	batches [][]string
	fail    error
	dim     int
}

func (e *outsideBatchEmbedder) Embed(texts []string) ([][]float64, error) {
	e.batches = append(e.batches, append([]string(nil), texts...))
	if e.fail != nil {
		return nil, e.fail
	}
	dim := e.dim
	if dim == 0 {
		dim = 3
	}
	out := make([][]float64, len(texts))
	for i, t := range texts {
		vec := make([]float64, dim)
		vec[0] = float64(len(t))
		out[i] = vec
	}
	return out, nil
}

type outsideSingleEmbedder struct {
	calls []string
	fail  error
}

func (e *outsideSingleEmbedder) Embed(text string) ([]float64, error) {
	e.calls = append(e.calls, text)
	if e.fail != nil {
		return nil, e.fail
	}
	return []float64{float64(len(text)), 1, 0}, nil
}

type outsideGenerator struct {
	seen []string
	fail error
}

func (g *outsideGenerator) Answer(question, passage, answerLanguage string) (string, error) {
	g.seen = append(g.seen, question+"|"+passage+"|"+answerLanguage)
	if g.fail != nil {
		return "", g.fail
	}
	return passage, nil
}

// The whole point, asserted at COMPILE time: shape alone is enough.
var (
	_ EmbeddingProvider  = (*outsideBatchEmbedder)(nil)
	_ SingleTextEmbedder = (*outsideSingleEmbedder)(nil)
	_ GeneratorProvider  = (*outsideGenerator)(nil)
)

// ---------------------------------------------------------------------------
// 1. The contracts exist and mean what Python's mean
// ---------------------------------------------------------------------------

func TestTheTwoConsumedSeamsArePublished(t *testing.T) {
	for name, iface := range map[string]reflect.Type{
		"EmbeddingProvider":  reflect.TypeOf((*EmbeddingProvider)(nil)).Elem(),
		"GeneratorProvider":  reflect.TypeOf((*GeneratorProvider)(nil)).Elem(),
		"SingleTextEmbedder": reflect.TypeOf((*SingleTextEmbedder)(nil)).Elem(),
	} {
		if iface.Kind() != reflect.Interface {
			t.Fatalf("%s must be an interface, got %v", name, iface.Kind())
		}
		if iface.NumMethod() != 1 {
			t.Errorf("%s has %d methods; one seam is one operation", name, iface.NumMethod())
		}
	}
}

// The three seams Python publishes that neither port consumes MUST NOT appear
// here. A contract with no call site advertises support that does not exist.
func TestTheUnconsumedSeamsAreAbsent(t *testing.T) {
	src := packageSource(t)
	for _, absent := range []string{
		"CompletionProvider", "VisionProvider", "RerankerProvider",
		"Candidate", "Complete(", "Describe(", "Rerank(",
	} {
		if strings.Contains(src, absent) {
			t.Errorf("the Go port publishes %q but has no consumer for it "+
				"(see design.md §1) — a contract nothing calls is worse than none", absent)
		}
	}
}

// Every contract returns (value, error). No sentinels: a provider that cannot
// fulfil a call must be able to SAY so (ADR-0014 R2).
func TestEveryContractMethodReturnsAnError(t *testing.T) {
	errType := reflect.TypeOf((*error)(nil)).Elem()
	for _, iface := range []reflect.Type{
		reflect.TypeOf((*EmbeddingProvider)(nil)).Elem(),
		reflect.TypeOf((*GeneratorProvider)(nil)).Elem(),
		reflect.TypeOf((*SingleTextEmbedder)(nil)).Elem(),
	} {
		m := iface.Method(0)
		if m.Type.NumOut() != 2 || m.Type.Out(1) != errType {
			t.Errorf("%s.%s must return (T, error); got %v", iface, m.Name, m.Type)
		}
	}
}

// R3: the contract names the OPERATION, never the transport. Asserted on both
// the parameter NAMES (from the AST) and the parameter TYPES (from reflect), so
// a provider that never opens a socket satisfies everything here.
func TestNoContractMentionsATransport(t *testing.T) {
	banned := map[string]bool{
		"baseurl": true, "baseUrl": true, "url": true, "headers": true,
		"transport": true, "timeout": true, "client": true, "apikey": true,
	}
	for iface, params := range interfaceParamNames(t) {
		for _, p := range params {
			if banned[strings.ToLower(p)] {
				t.Errorf("%s takes a transport concern %q — R3 forbids it", iface, p)
			}
		}
	}

	for _, iface := range []reflect.Type{
		reflect.TypeOf((*EmbeddingProvider)(nil)).Elem(),
		reflect.TypeOf((*GeneratorProvider)(nil)).Elem(),
		reflect.TypeOf((*SingleTextEmbedder)(nil)).Elem(),
	} {
		mt := iface.Method(0).Type
		for i := 0; i < mt.NumIn(); i++ {
			switch mt.In(i).Kind() {
			case reflect.Map, reflect.Func, reflect.Chan, reflect.Interface:
				t.Errorf("%s parameter %d is a %v — that is a transport shape, not an operation",
					iface, i, mt.In(i).Kind())
			}
		}
	}
}

// A provider author's only dependency must be this file.
func TestTheContractsPackageDependsOnNothingOfOurs(t *testing.T) {
	fset := token.NewFileSet()
	pkgs, err := parser.ParseDir(fset, ".", func(fi os.FileInfo) bool {
		return !strings.HasSuffix(fi.Name(), "_test.go")
	}, 0)
	if err != nil {
		t.Fatalf("parse package: %v", err)
	}
	for _, pkg := range pkgs {
		for name, file := range pkg.Files {
			for _, imp := range file.Imports {
				if strings.Contains(imp.Path.Value, "citenexus") {
					t.Errorf("%s imports %s; the contracts package must import nothing of ours",
						filepath.Base(name), imp.Path.Value)
				}
			}
		}
	}
}

// ---------------------------------------------------------------------------
// 2. The two embedding shapes cannot be confused — the Python hazard, checked
// ---------------------------------------------------------------------------

// Python had to name the batch method `embed_many` because `str` IS a
// `Sequence[str]`, so `embed` could not be told apart from the single-text
// shape. In Go `string` and `[]string` are unrelated types, so the natural name
// is safe — and this test is the proof, not the claim.
func TestTheSingleAndBatchShapesAreNotInterchangeable(t *testing.T) {
	var single any = &outsideSingleEmbedder{}
	if _, ok := single.(EmbeddingProvider); ok {
		t.Error("a single-text embedder satisfied the batch contract — the Python hazard IS present in Go")
	}
	var batch any = &outsideBatchEmbedder{}
	if _, ok := batch.(SingleTextEmbedder); ok {
		t.Error("a batch provider satisfied the single-text contract")
	}
}

// ---------------------------------------------------------------------------
// 3. Dispatch: batch preferred, single-text fallback, order preserved
// ---------------------------------------------------------------------------

func TestEmbedTextsPrefersTheBatchContract(t *testing.T) {
	batch := &outsideBatchEmbedder{}
	vecs, err := EmbedTexts(batch, []string{"a", "bb", "ccc"})
	if err != nil {
		t.Fatalf("EmbedTexts: %v", err)
	}
	if len(batch.batches) != 1 {
		t.Fatalf("batch path not taken once: %v", batch.batches)
	}
	if len(vecs) != 3 {
		t.Fatalf("got %d vectors for 3 texts", len(vecs))
	}
	// Order preserved: the fake encodes len(text) in slot 0.
	for i, want := range []float64{1, 2, 3} {
		if vecs[i][0] != want {
			t.Errorf("vector %d is out of input order: %v", i, vecs[i])
		}
	}
}

func TestEmbedTextsFallsBackToTheSingleTextShape(t *testing.T) {
	single := &outsideSingleEmbedder{}
	vecs, err := EmbedTexts(single, []string{"a", "bb"})
	if err != nil {
		t.Fatalf("EmbedTexts: %v", err)
	}
	if len(vecs) != 2 || len(single.calls) != 2 {
		t.Fatalf("single-text fallback wrong: vecs=%d calls=%v", len(vecs), single.calls)
	}
	if single.calls[0] != "a" || single.calls[1] != "bb" {
		t.Errorf("input order lost: %v", single.calls)
	}
}

func TestEmbedTextsOnEmptyInputCallsNothing(t *testing.T) {
	batch := &outsideBatchEmbedder{}
	vecs, err := EmbedTexts(batch, nil)
	if err != nil || len(vecs) != 0 {
		t.Fatalf("empty input: vecs=%v err=%v", vecs, err)
	}
	if len(batch.batches) != 0 {
		t.Errorf("an empty batch was still sent to the model: %v", batch.batches)
	}
}

var errBoom = errors.New("model call timed out after 30s")

func TestEmbedTextsPropagatesTheProviderError(t *testing.T) {
	if _, err := EmbedTexts(&outsideBatchEmbedder{fail: errBoom}, []string{"a"}); !errors.Is(err, errBoom) {
		t.Fatalf("batch error swallowed: %v", err)
	}
	if _, err := EmbedTexts(&outsideSingleEmbedder{fail: errBoom}, []string{"a"}); !errors.Is(err, errBoom) {
		t.Fatalf("single-text error swallowed: %v", err)
	}
}

func TestEmbedTextsRefusesSomethingThatIsNeither(t *testing.T) {
	_, err := EmbedTexts(struct{}{}, []string{"a"})
	if !errors.Is(err, ErrNotAnEmbedder) {
		t.Fatalf("a non-embedder was accepted: %v", err)
	}
	if _, err := EmbedTexts(nil, []string{"a"}); !errors.Is(err, ErrNotAnEmbedder) {
		t.Fatalf("nil was accepted: %v", err)
	}
}

func TestEmbedOneIsABatchOfOne(t *testing.T) {
	batch := &outsideBatchEmbedder{}
	vec, err := EmbedOne(batch, "hello")
	if err != nil {
		t.Fatalf("EmbedOne: %v", err)
	}
	if vec[0] != 5 {
		t.Errorf("wrong vector: %v", vec)
	}
	if len(batch.batches) != 1 || len(batch.batches[0]) != 1 {
		t.Errorf("a single text must be a batch of one, got %v", batch.batches)
	}
}

// ---------------------------------------------------------------------------
// 4. SingleFrom — a batch provider plugs into the single-text ingest seam
// ---------------------------------------------------------------------------

func TestSingleFromAdaptsABatchProvider(t *testing.T) {
	batch := &outsideBatchEmbedder{}
	var adapted SingleTextEmbedder = SingleFrom(batch)
	vec, err := adapted.Embed("hello")
	if err != nil {
		t.Fatalf("SingleFrom: %v", err)
	}
	if vec[0] != 5 {
		t.Errorf("wrong vector: %v", vec)
	}
}

func TestSingleFromReportsAShortBatchRatherThanAPlaceholder(t *testing.T) {
	_, err := SingleFrom(shortBatch{}).Embed("hello")
	if err == nil {
		t.Fatal("a provider that returned no vector was reported as success")
	}
}

type shortBatch struct{}

func (shortBatch) Embed(texts []string) ([][]float64, error) { return nil, nil }

// ---------------------------------------------------------------------------
// 5. The shared vector guard
// ---------------------------------------------------------------------------

func TestCheckVector(t *testing.T) {
	for _, c := range []struct {
		name string
		vec  []float64
		dim  int
		want string // substring the error must name; "" = must be accepted
	}{
		{"zero", []float64{0, 0, 0, 0}, 4, "zero vector"},
		{"nil", nil, 0, "empty vector"},
		{"empty", []float64{}, 0, "empty vector"},
		{"wrong dim", []float64{1, 2}, 4, "2-dim vector"},
		{"defines the run", []float64{0, 0, 0.5, 0}, 0, ""},
		{"matching dim", []float64{0, 0, 0.5, 0}, 4, ""},
	} {
		err := CheckVector("eu", c.vec, c.dim)
		if c.want == "" {
			if err != nil {
				t.Errorf("%s: rejected a valid vector: %v", c.name, err)
			}
			continue
		}
		if err == nil {
			t.Errorf("%s: accepted a vector that must never be indexed", c.name)
		} else if !strings.Contains(err.Error(), c.want) {
			t.Errorf("%s: error should name %q, got %v", c.name, c.want, err)
		}
	}
}

func TestIsZeroVector(t *testing.T) {
	for _, c := range []struct {
		vec  []float64
		want bool
	}{
		{[]float64{0, 0, 0}, true},
		{[]float64{}, true},
		{[]float64{0, 0, -0.0}, true},
		{[]float64{0, 1e-12, 0}, false},
		{[]float64{-1, 0, 0}, false},
	} {
		if got := IsZeroVector(c.vec); got != c.want {
			t.Errorf("IsZeroVector(%v) = %v, want %v", c.vec, got, c.want)
		}
	}
}

// ---------------------------------------------------------------------------
// helpers
// ---------------------------------------------------------------------------

func packageSource(t *testing.T) string {
	t.Helper()
	entries, err := os.ReadDir(".")
	if err != nil {
		t.Fatalf("read package dir: %v", err)
	}
	var sb strings.Builder
	for _, e := range entries {
		if !strings.HasSuffix(e.Name(), ".go") || strings.HasSuffix(e.Name(), "_test.go") {
			continue
		}
		raw, err := os.ReadFile(e.Name())
		if err != nil {
			t.Fatalf("read %s: %v", e.Name(), err)
		}
		sb.Write(raw)
	}
	return sb.String()
}

// interfaceParamNames maps each exported interface in the package to the
// parameter names of its methods, read from the AST (reflect drops names).
func interfaceParamNames(t *testing.T) map[string][]string {
	t.Helper()
	fset := token.NewFileSet()
	pkgs, err := parser.ParseDir(fset, ".", func(fi os.FileInfo) bool {
		return !strings.HasSuffix(fi.Name(), "_test.go")
	}, 0)
	if err != nil {
		t.Fatalf("parse package: %v", err)
	}
	out := map[string][]string{}
	for _, pkg := range pkgs {
		for _, file := range pkg.Files {
			ast.Inspect(file, func(n ast.Node) bool {
				ts, ok := n.(*ast.TypeSpec)
				if !ok {
					return true
				}
				it, ok := ts.Type.(*ast.InterfaceType)
				if !ok {
					return true
				}
				for _, m := range it.Methods.List {
					ft, ok := m.Type.(*ast.FuncType)
					if !ok || ft.Params == nil {
						continue
					}
					for _, p := range ft.Params.List {
						for _, id := range p.Names {
							out[ts.Name.Name] = append(out[ts.Name.Name], id.Name)
						}
					}
				}
				return true
			})
		}
	}
	if len(out) == 0 {
		t.Fatal("no interfaces found — the contracts package is empty")
	}
	return out
}
